schema.cc

```diff
@@ -1618,6 +1618,17 @@ schema_ptr schema::get_reversed() const {
     });
 }
 
+void schema::set_gc_before(std::function<gc_clock::time_point (const dht::decorated_key& dk, const gc_clock::time_point& query_time)> func) const {
+    *_get_gc_before = std::move(func);
```

Comment:
The schema object is immutable by design, it should not have setters like this.

Also, serializing and deserializing into schema mutations should give the same object. If we rely on setters, this will not work. `gobal_schema_ptr` will lose _get_gc_before when you move it across shards.

Reply:
I know we are not supposed to modify it by the current design, the schema_ptr is seastar::lw_shared_ptr<const schema>.  I am looking possibilities to change this a bit.  I do not want to mess up with schema at all. But there are so many places where only a schema is available while we need to call get_gc_before(). I currently do not find a  better solution than make get_gc_before() available to schema object.  If there is a better way, I am more than happy to use it. 

The set_gc_before is called when a schema is attached with a specific table. Will all the places start to use the schmea_ptr attached to a table eventually? If this is not the case, calling set_gc_before at table::set_schema is not enough. 

@avikivity @tgrabiec  This is a major challenge and a blocker. Let's find a good solution. What do you suggest?

Reply:
IMO mutating the schema is asking for major trouble. What are the places where you need gc_before? It's really only compaction and mutation_partition::compact_for_query(). These are called from relatively top-level places that should have access to repair state.

Reply:
It could be that the read is using a schema which comes from the registry and is different than the current table's schema (different version). In general, it seems unsafe to rely on this.

Would it be difficult to propagate the gc_before function?

Reply:
Yes, it is very hard to propagate gc_before function everywhere.

Here below is part of them I looked at. 

```
- result_set::result_set
  calls auto qr = query_mutation(mutation(m), slice)
  which calls mutation_partition.cc::query_mutation
  which calls compact_for_query<>
  used by db/schema_tables.cc, schema_mutations.cc

- mutation_partition.cc::to_data_query_result  (All calls have the db() object)
  being used in storage_proxy.cc
  uses compact_for_query<>

- mutation_partition.cc::query_mutation
  uses compact_for_query<>
  Used by result_set::result_set

- mutation_partition.cc::counter_write_query
  uses compact_for_query<>

- mutation_partition.cc::do_compact

- db/view/view.cc::view_builder::execute
  compact_for_query<>
```

Reply:
However, I think I found a much better solution.  Let me explain. 

I moved out the repair history map from the table object. So, the get_gc_before function does not need the table object anymore. This is a major simplification. 

I put a shared pointer to the repair history map inside  a schema whenever the schema  is constructed. 

See the last patch in version v6. 

Like the _partitioner and sharder, _repair_history_map is not stored in the schema mutation. It is just a helper to get the repair history map. 

```
struct raw_schema {
... 
 664         std::reference_wrapper<const dht::i_partitioner> _partitioner;
 665         // Sharding info is not stored in the schema mutation and does not affect
 666         // schema digest. It is also not set locally on a schema tables.
 667         std::reference_wrapper<const dht::sharder> sharder;
 668         std::shared_ptr<repair_history_map> _repair_history_map;
}
```

The _repair_history_map points to a per shard per table repair_history_map object. 

We do not set anything to the schema object. The schema remains const. 

@avikivity @tgrabiec What do you think?

Reply:
std::shared_ptr? That's asking for a different kind of trouble.

Reply:
especially as repair_history_map is mutable.  And how can you be sure every schema gets a repair_history_map?

Reply:
Also, how will freezing the schema on one shard and unfreezing it on another one (on another node possibly) work?

Reply:
> Yes, it is very hard to propagate gc_before function everywhere.
> 
> Here below is part of them I looked at.
> 
> ```
> - result_set::result_set
>   calls auto qr = query_mutation(mutation(m), slice)
>   which calls mutation_partition.cc::query_mutation
>   which calls compact_for_query<>
>   used by db/schema_tables.cc, schema_mutations.cc

These are local tables that can gc immediately.


> 
> - mutation_partition.cc::to_data_query_result  (All calls have the db() object)
>   being used in storage_proxy.cc
>   uses compact_for_query<>
> 

storage_proxy has access to the table object.

> - mutation_partition.cc::query_mutation
>   uses compact_for_query<>
>   Used by result_set::result_set
> 

already described above.

> - mutation_partition.cc::counter_write_query
>   uses compact_for_query<>
> 

counter_write_query is called from storage_proxy, which should have access to the table object.

> - mutation_partition.cc::do_compact
> 

Who's the real caller?

> - db/view/view.cc::view_builder::execute


Should have access to the table object.

>   compact_for_query<>
> ```

Reply:
I am using std::shared_ptr instead of lw_shared_ptr because the latter does not allow me to use with incomplete types in the header.

Reply:
Whenever the raw_schema object is constructed, it is assigned a repair_history_map shared pointer. Is this enough?

Reply:
Like the reference we attached to raw_schema 

std::reference_wrapper<const dht::i_partitioner> _partitioner;
std::reference_wrapper<const dht::sharder> _sharder;

repair_history_map is local to the shard and also local to the node itself.

I think frozen and unfrozen will not have impact at all.

Reply:
> I am using std::shared_ptr instead of lw_shared_ptr because the latter does not allow me to use with incomplete types in the header.


seastar::shared_ptr suppose to work for that as well.

Reply:
mutation_partition::compact_for_query and mutation_partition::compact_for_compaction call do_compact().